// dt/lib/app.js
/*jslint node: true, vars: true */
'use strict';

//Copyright 2014 Resilient Network Systems, Inc.
//
//Licensed under the Apache License, Version 2.0 (the "License");  
//you may not use this file except in compliance with the License.  
//You may obtain a copy of the License at  
//
//     http://www.apache.org/licenses/LICENSE-2.0  
//
//Unless required by applicable law or agreed to in writing, software  
//distributed under the License is distributed on an "AS IS" BASIS,  
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
//See the License for the specific language governing permissions and  
//limitations under the License. 

//#Decision Table Processing

//**compile** - compile a decision table
//
//+ **dt** - a decision table in JSON LD format; see dt_types.js for details
//+ **returns** an anonymous function that implements the decision table
//
//**trace**
//
//+ **dt** - same as compile
//+ **returns** - this function is the same as from compile; however, 
//it must be given an extra last parameter that is the log to which 
//it will trace decision table logic; the log must be an array; the 
//function appends to it string trace messages
//
//**import_spreadsheet** - not implemented
//
//**export_spreadsheet** - not implemented
//
//**dt_t** - decision table JSON LD type identifiers
//
//**dt_p** - decision table JSON LD property identifiers

module.exports = {
    compile: require('./dt_compile').compile,
    trace: require('./dt_trace').trace,
    import_spreadsheet: require('./dt_import').import_spreadsheet,
    export_spreadsheet: require('./dt_export').export_spreadsheet,
    dt_t: require('./dt_common').dt_t,
    dt_p: require('./dt_common').dt_p
};
