// dt/lib/dt_common.js
/*jslint node: true, vars: true */
'use strict';

//Copyright 2014 Resilient Network Systems, Inc.
//
//Licensed under the Apache License, Version 2.0 (the "License");  
//you may not use this file except in compliance with the License.  
//You may obtain a copy of the License at  
//
//     http://www.apache.org/licenses/LICENSE-2.0  
//
//Unless required by applicable law or agreed to in writing, software  
//distributed under the License is distributed on an "AS IS" BASIS,  
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
//See the License for the specific language governing permissions and  
//limitations under the License. 

//#Common DT Constants

var dt_t = {
    dt: "http://type.dt.tbd.org/dt",
    balanced_dt: "http://type.dt.tbd.org/balanced_dt",
    first_dt: "http://type.dt.tbd.org/first_dt",
    all_dt: "http://type.dt.tbd.org/all_dt",
    json: "http://type.dt.tbd.org/json",
    dt_name: "http://type.dt.tbd.org/dt_name",
    inner_dt_name: "http://type.dt.tbd.org/inner_dt_name",
    for_loop: "http://type.dt.tbd.org/for_loop",
    policy: "http://type.dt.tbd.org/policy"
};

var dt_p = {
    inner_dts: "http://property.dt.tbd.org/inner_dts",
    dt_name: "http://property.dt.tbd.org/dt_name",
    comments: "http://property.dt.tbd.org/comments",
    parameter_descriptions: "http://property.dt.tbd.org/parameter_descriptions",
    parameters: "http://property.dt.tbd.org/parameters",
    variable_descriptions: "http://property.dt.tbd.org/variable_descriptions",
    variables: "http://property.dt.tbd.org/variables",
    initial_action_descriptions: "http://property.dt.tbd.org/initial_action_descriptions",
    initial_actions: "http://property.dt.tbd.org/initial_actions",
    for_descriptions: "http://property.dt.tbd.org/for_descriptions",
    for_loops: "http://property.dt.tbd.org/for_loops",
    for_element: "http://property.dt.tbd.org/for_element",
    for_array: "http://property.dt.tbd.org/for_array",
    for_where: "http://property.dt.tbd.org/for_where",
    for_loop_initial_action_descriptions: "http://property.dt.tbd.org/for_loop_initial_action_descriptions",
    for_loop_initial_actions: "http://property.dt.tbd.org/for_loop_initial_actions",
    condition_descriptions: "http://property.dt.tbd.org/condition_descriptions",
    conditions: "http://property.dt.tbd.org/conditions",
    action_descriptions: "http://property.dt.tbd.org/action_descriptions",
    actions: "http://property.dt.tbd.org/actions",
    policy_descriptions: "http://property.dt.tbd.org/policy_descriptions",
    policies: "http://property.dt.tbd.org/policies",
    condition_pattern: "http://property.dt.tbd.org/condition_pattern",
    action_selector: "http://property.dt.tbd.org/action_selector",
    executor: "http://property.dt.tbd.org/executor"
};

module.exports = {
    dt_t: dt_t,
    dt_p: dt_p
};
