// roster/roster.js
/*jslint node: true, vars: true */
'use strict';

//This is part of the roster test for the dt module
//
var r_t = {
    roster_input: "http://type.dt.example.edu/roster_input",
    student: "http://type.dt.example.edu/student",
    name: "http://type.dt.example.edu/name",
    age: "http://type.dt.example.edu/age",
    race: "http://type.dt.example.edu/race",
    school: "http://type.dt.example.edu/school"
};

var r_p = {
    county_name: "http://property.dt.example.edu/county_name",
    date: "http://property.dt.example.edu/date",
    students: "http://property.dt.example.edu/students",
    name: "http://property.dt.example.edu/name",
    age: "http://property.dt.example.edu/age",
    race: "http://property.dt.example.edu/race",
    school: "http://property.dt.example.edu/school"
};

function addCounty(name, roster) {
    var rosterEntry;

    rosterEntry = 'County: ' + name;
    roster.push(rosterEntry);
}

function add(name, rosterCategory, roster) {
    var rosterEntry;

    switch (rosterCategory) {
    case 'roster':
        rosterEntry = 'roster: ' + name;
        break;
    case 'twelve-and-under':
        rosterEntry = 'twelve-and-under: ' + name;
        break;
    case 'PS105-student':
        rosterEntry = 'PS105-student: ' + name;
        break;
    case 'PS106-student':
        rosterEntry = 'PS106-student: ' + name;
        break;
    case 'caucasion-student':
        rosterEntry = 'caucasion-student: ' + name;
        break;
    case 'other-student':
        rosterEntry = 'other-student: ' + name;
        break;
    default:
        rosterEntry = 'bad roster category';
    }

    roster.push(rosterEntry);

} //add

module.exports = {
    r_t: r_t,
    r_p: r_p,
    addCounty: addCounty,
    add: add
};
