// roster/roster.js
/*jslint node: true, vars: true */
'use strict';

//Copyright 2014 Resilient Network Systems, Inc.
//
//Licensed under the Apache License, Version 2.0 (the "License");  
//you may not use this file except in compliance with the License.  
//You may obtain a copy of the License at  
//
//     http://www.apache.org/licenses/LICENSE-2.0  
//
//Unless required by applicable law or agreed to in writing, software  
//distributed under the License is distributed on an "AS IS" BASIS,  
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  
//See the License for the specific language governing permissions and  
//limitations under the License. 

//This is part of the roster test for the dt module
//
var r_t = {
    roster_input: "http://type.dt.example.edu/roster_input",
    student: "http://type.dt.example.edu/student",
    name: "http://type.dt.example.edu/name",
    age: "http://type.dt.example.edu/age",
    race: "http://type.dt.example.edu/race",
    school: "http://type.dt.example.edu/school"
};

var r_p = {
    county_name: "http://property.dt.example.edu/county_name",
    date: "http://property.dt.example.edu/date",
    students: "http://property.dt.example.edu/students",
    name: "http://property.dt.example.edu/name",
    age: "http://property.dt.example.edu/age",
    race: "http://property.dt.example.edu/race",
    school: "http://property.dt.example.edu/school"
};

function addCounty(name, roster) {
    var rosterEntry;

    rosterEntry = 'County: ' + name;
    roster.push(rosterEntry);
}

function add(name, rosterCategory, roster) {
    var rosterEntry;

    switch (rosterCategory) {
    case 'roster':
        rosterEntry = 'roster: ' + name;
        break;
    case 'twelve-and-under':
        rosterEntry = 'twelve-and-under: ' + name;
        break;
    case 'PS105-student':
        rosterEntry = 'PS105-student: ' + name;
        break;
    case 'PS106-student':
        rosterEntry = 'PS106-student: ' + name;
        break;
    case 'caucasion-student':
        rosterEntry = 'caucasion-student: ' + name;
        break;
    case 'other-student':
        rosterEntry = 'other-student: ' + name;
        break;
    default:
        rosterEntry = 'bad roster category';
    }

    roster.push(rosterEntry);

} //add

module.exports = {
    r_t: r_t,
    r_p: r_p,
    addCounty: addCounty,
    add: add
};
